server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1190,18 +1198,30 @@ private void finalizeSnapshotEntry(SnapshotsInProgress.Entry entry, Metadata met
             } else {
                 metadataListener.onResponse(metadata);
             }
-            metadataListener.whenComplete(meta -> repo.finalizeSnapshot(
-                    shardGenerations,
-                    repositoryData.getGenId(),
-                    metadataForSnapshot(entry, meta),
-                    snapshotInfo,
-                    entry.version(),
-                    state -> stateWithoutSnapshot(state, snapshot),
-                    ActionListener.wrap(newRepoData -> {
-                        completeListenersIgnoringException(endAndGetListenersToResolve(snapshot), Tuple.tuple(newRepoData, snapshotInfo));
-                        logger.info("snapshot [{}] completed with state [{}]", snapshot, snapshotInfo.state());
-                        runNextQueuedOperation(newRepoData, repository, true);
-                    }, e -> handleFinalizationFailure(e, entry, repositoryData))),
+            metadataListener.whenComplete(meta -> {
+                        final Metadata metaForSnapshot = metadataForSnapshot(entry, meta);
+                        final SnapshotInfo snapshotInfo = new SnapshotInfo(snapshot.getSnapshotId(),
```

Comment:
So by building `SnapshotInfo` at a later stage when it is know what has been snapshotted we avoid including data streams in a snapshot that refer to indices that don't exist anymore? (to protect against creating a snapshot and at the same time deleting a backing index?)

Reply:
Yea exactly. If we work out the metadata first and then build `SnapshotInfo` then we can use it to compute the datastreams that actually made it into the snapshot without doing extra work.